File: module-services/src/main/java/eu/geekhome/services/automation/EvaluationResult.java
package eu.geekhome.services.automation;

import eu.geekhome.services.localization.Resource;

import java.util.HashMap;
import java.util.Map;

public class EvaluationResult {

    private Object _value;
    private Resource _interfaceValue;
    private Map<String, Resource> _descriptions;
    private ControlMode _controlMode;
    private boolean _isSignaled;
    private String _colorValue;

    public Object getValue() {
        return _value;
    }

    public void setValue(Object value) {
        _value = value;
    }

    public boolean isSignaled() {
        return _isSignaled;
    }

    public void setSignaled(boolean value) {
        _isSignaled = value;
    }

    public Resource getInterfaceValue() {
        return _interfaceValue;
    }

    public String getColorValue() {
        return _colorValue;
    }

    public void setColorValue(String colorValue) {
        _colorValue = colorValue;
    }

    public void setInterfaceValue(Resource value) {
        _interfaceValue = value;
    }

    public Map<String, Resource> getDescriptions() {
        return _descriptions;
    }

    public void setControlMode(ControlMode controlMode) {
        _controlMode = controlMode;
    }

    public ControlMode getControlMode() {
        return _controlMode;
    }

    public void setDescriptions(Map<String, Resource> value) {
        _descriptions = value;
    }

    public EvaluationResult(Object value, Resource interfaceValue, boolean isSignaled,
                            Map<String, Resource> descriptions, ControlMode controlMode) {
        setDescriptions(descriptions);
        setValue(value);
        setInterfaceValue(interfaceValue);
        setControlMode(controlMode);
        setSignaled(isSignaled);
    }

    public EvaluationResult(Object value, Resource interfaceValue, boolean isSignaled,
                            Map<String, Resource> descriptions) {
        this(value, interfaceValue, isSignaled, descriptions, ControlMode.Auto);
    }

    public EvaluationResult(Object value, Resource interfaceValue, boolean isSignaled) {
        this(value, interfaceValue, isSignaled, new HashMap<>());
    }
}

File: module-services/src/main/java/eu/geekhome/services/automation/EvaluationResult.kt
package eu.geekhome.services.automation

import eu.geekhome.services.localization.Resource

class EvaluationResult<T> (
    val interfaceValue: Resource?,
    val value: T? = null,
    val isSignaled: Boolean = false,
    val descriptions: Map<String, Resource> = HashMap(),
    val controlMode: ControlMode = ControlMode.Auto,
    val error: Exception? = null,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as EvaluationResult<*>

        if (interfaceValue != other.interfaceValue) return false
        if (value != other.value) return false
        if (isSignaled != other.isSignaled) return false
        if (descriptions != other.descriptions) return false
        if (controlMode != other.controlMode) return false
        if (error != other.error) return false

        return true
    }

    override fun hashCode(): Int {
        var result = interfaceValue?.hashCode() ?: 0
        result = 31 * result + (value?.hashCode() ?: 0)
        result = 31 * result + isSignaled.hashCode()
        result = 31 * result + descriptions.hashCode()
        result = 31 * result + controlMode.hashCode()
        result = 31 * result + (error?.hashCode() ?: 0)
        return result
    }
}